### ISSUE TITLE
Lagging experience

### ISSUE THREAD

**Issue by Wordluc** [seblj/roslyn.nvim#93]
Even with filewatching set to false, the LSP server process still consumes up to 50% of CPU usage, which suggests it might still be monitoring the file, causing the entire experience to lag. In contrast, when using Visual Studio Code with the same file, the CPU usage does not increase.

**Comment by seblj** [comment]
Yes I have also noticed that it has been slower recently... From what I can see from the code I have, it should still do what it did before I removed a hack for something the server didn't follow the spec for...

I can try to investigate when I get some time

**Comment by Wordluc** [comment]
i can try to use a old version of the server and the client, to see.

**Comment by seblj** [comment]
Thanks! Could you also try old version of the plugin before I removed hacks.lua?

**Comment by Wordluc** [comment]
i moved here '[fix: fix bug with sln parser](https://github.com/seblj/roslyn.nvim/commit/bfca961e851d0ba1bf759856051563a788ee4851)', commit before '[fix: remove unnecessary hack for filewatching](https://github.com/seblj/roslyn.nvim/commit/a2f3b629761e79de21e0ae0a35cdab628410d95b)' and the problem is still there

**Comment by Wordluc** [comment]
and i wasnt able to use an old version of roslyn it gives me a problem.I feel the is something with the new version of roslyn but i wasnt able to test it

**Comment by mparq** [comment]
Came to the issues because I've been noticing the same thing recently w/ roslyn and rzls. Shamelessly plugging a small go program I just wrote for myself to help understand the lsp flow for developing on rzls.nvim. https://github.com/mparq/lsptrace in case the lsp traces can reveal anything strange that might be causing the lag.

**Comment by seblj** [comment]
> Came to the issues because I've been noticing the same thing recently w/ roslyn and rzls. Shamelessly plugging a small go program I just wrote for myself to help understand the lsp flow for developing on rzls.nvim. https://github.com/mparq/lsptrace in case the lsp traces can reveal anything strange that might be causing the lag.

Very interesting! I will see if I can try it out. Feel free, anyone, to comment here about any potential findings. Any help would be more than welcomed!

**Comment by Wordluc** [comment]
i found issue of vscode's extention(since it uses roslyn too) that look like our problem: https://github.com/dotnet/vscode-csharp/issues/4350

**Comment by seblj** [comment]
I am not sure if that is the exact problem🤔 Because I feel like it has gotten super slow recently without my project necessarily growing a lot in size.

Today it has been unbearable at some point, where I needed to restart neovim and it just used all of mye CPU and my fan on my macbook pro m2 pro just started spinning up for like the first time ever

**Comment by drzbida** [comment]
is there any info you could give for when it's slow? what is exactly slow: code editing / navigation / renferences, etc. how does it become slow, are there random spikes that make it suddenly slow, or is it gradually becoming slower over time? if so how much time does it usually take, how many buffers do you have open, etc ? anything could probably help 

i've tried to reproduce this quite a bit today, but without much success. the only thing weird for me was that somehow trying to type vim commands became abnormally slow, but somehow code editing and navigation worked normally

**Comment by seblj** [comment]
Yeah sure!

So, it seemed to be somewhat random for me today, but it happened multiple times. I don't really have solid reproducible steps, but it seemed to happen after having edited some files for about 1 hour or so. I know I had multiple buffers open, but I am not sure exactly how many. If I were to guess I would say between 5 and 10 open buffers, but most of them just in the background

It seemed to just suddenly become slow, and then everything would basically choke. It would take several seconds before being able to type something at all, and for text to come up on the screen. Also, diagnostics were super slow.

Trying to do anything like go to definition, hover and navigation would just not work, and I would have to just wait for several seconds before it would start to slow down again. However, if I remember correctly, it seemed like it would suddenly spike again without doing too much after it had happened once.

Restarting neovim seemed to make it a lot better and then I would be able to use it for max an hour or so before it started happening again.

I looked at my CPU usage in htop, and all my cores was maxed out at 100% with roslyn taking 700-800% of CPU

**Comment by Wordluc** [comment]
in my case, it happend when i am in a big file(>1000 lines),when i try to edit it the lsp takes >10 seconds to update the warnings

**Comment by seblj** [comment]
Do you have a consistent reproduction? If so, I have small suspicion that some of this might be related to diagnostics.

So I am setting dynamicregistrations for diagnostics as a capability even though neovim doesn't support it. I haven't looked too much into what it means, but roslyn doesn't send any diagnostics if not. You can see in config.lua

I know that roslyn uses pull diagnostics, which means neovim is asking for diagnostics, and from what I can see, this happens on every single change.

Could it maybe be that roslyn is not discarding any previous requests for diagnostics, even though new ones are coming in? And then it needs to handle a lot at once to send diagnostics for old buffer state?

If someone has a good reproduction for this, it would be really helpful if you could share it. For example by adding a link to a project here on github and such.

If not possible, could someone for example try to disable the diagnostics by modifying the config.lua file and/or try something like this out: https://github.com/neovim/neovim/pull/30055/files

I have no idea if some of that would help, or even is the issue here, but this is what I am thinking right now

**Comment by Wordluc** [comment]
i think you got it, when i write "var a = "foo"" , i got the warning in different steps:
1)i got "v" dosnt exist
2) got "var" error....
3) "got var a = ..."
It seems that is buffering the warning

**Comment by Wordluc** [comment]
i've just tried to disable the diagnostic "vim.diagnostic.enable(false)", but still the cpu go up

**Comment by Wordluc** [comment]
with this two 
```
	 roslyn_config.config.capabilities.textDocument.diagnostic.dynamicRegistration=true
	 roslyn_config.config.capabilities.textDocument.diagnostic.relatedDocumentSupport=true
```
it seems that something is speeding up(i dont know if it's my impression)

**Comment by drzbida** [comment]
> 
> I know that roslyn uses pull diagnostics, which means neovim is asking for diagnostics, and from what I can see, this happens on every single change.


I shouldn't send requests on every change, but only after 150ms after you've stopped typing (default value for `debounce_text_changes`) 

> 
> Could it maybe be that roslyn is not discarding any previous requests for diagnostics, even though new ones are coming in? And then it needs to handle a lot at once to send diagnostics for old buffer state?

Yes, that's true. It doesn't discard any request it doesn't receive a cancelRequest for. But both of those are pretty much true for every other LSP. I'll try a bit later to generate a few tens of thousands lines c# file to see if it becomes reproductible for me.

By the way, are the both of you using mac? I've recommended this plugin to someone using mac a few days ago and it just couldn't work because of some antivirus issues, maybe it's related somehow... My pc is almost definitely slower than a macbook pro m2 pro and it's very hard to reproduce on Windows.

**Comment by mparq** [comment]
I'm probably only confirming what has been already found but in case it's helpful, the trace of what I'm seeing between nvim and vscode (lines truncated to 200 chars) when I try to open the roslyn c# project itself.

it seems like vscode-csharp also sends off a bunch of diagnostic requests per keystroke, but the textDocument/diagnostic requests have extra information like `identifier` `previousResultId`, etc... whereas in nvim the requests don't have that. for nvim, the server responds after seconds whereas in vscode the responses are much more immediate.

nvim (see `**FIRST DIAGNOSTIC**` and `**RESPONSE TO FIRST DIAGNOSTIC**`):
```
NTF--> [textDocument/didChange] {"contentChanges":[{"text":"p","range":{"start":{"character":8,"line":40},"end":{"character":8,"line":40}},"rangeLength":0}],"textDocument":{"version":5,"uri":"file:\/\
REQ--> [textDocument/completion] id=25  {"context":{"triggerKind":1},"position":{"character":9,"line":40},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/CSharp\/Por
**FIRST DIAGNOSTIC**REQ--> [textDocument/diagnostic] id=26  {"range":{"start":{"line":5,"character":0},"end":{"line":49,"character":0}},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/C
NTF--> [textDocument/didChange] {"contentChanges":[{"text":"r","range":{"start":{"character":9,"line":40},"end":{"character":9,"line":40}},"rangeLength":0}],"textDocument":{"version":6,"uri":"file:\/\
REQ--> [textDocument/diagnostic] id=27  {"range":{"start":{"line":5,"character":0},"end":{"line":49,"character":0}},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/C
NTF--> [textDocument/didChange] {"contentChanges":[{"text":"i","range":{"start":{"character":10,"line":40},"end":{"character":10,"line":40}},"rangeLength":0}],"textDocument":{"version":7,"uri":"file:\
REQ--> [textDocument/diagnostic] id=28  {"range":{"start":{"line":5,"character":0},"end":{"line":49,"character":0}},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/C
<--RSP 315ms [textDocument/completion] id=25  {"isIncomplete":true,"itemDefaults":{"data":{"TextDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/CSharp\/Portable\/Compiler
REQ--> [textDocument/completion] id=29  {"context":{"triggerKind":3},"position":{"character":11,"line":40},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/CSharp\/Po
NTF--> [textDocument/didChange] {"contentChanges":[{"text":"v","range":{"start":{"character":11,"line":40},"end":{"character":11,"line":40}},"rangeLength":0}],"textDocument":{"version":8,"uri":"file:\
REQ--> [completionItem/resolve] id=30  {"insertTextFormat":1,"textEdit":{"newText":"private","range":{"start":{"character":8,"line":40},"end":{"character":9,"line":40}}},"kind":14,"data":{"ResultId":0
REQ--> [textDocument/diagnostic] id=31  {"range":{"start":{"line":5,"character":0},"end":{"line":49,"character":0}},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/C
<--RSP 36ms [completionItem/resolve] id=30  {"insertTextFormat":1,"data":{"ResultId":0,"TextDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/CSharp\/Portable\/Compiler\/Me
<--RSP 168ms [textDocument/completion] id=29  {"isIncomplete":true,"itemDefaults":{"data":{"TextDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/CSharp\/Portable\/Compiler
NTF--> [textDocument/didChange] {"contentChanges":[{"text":"a","range":{"start":{"character":12,"line":40},"end":{"character":12,"line":40}},"rangeLength":0}],"textDocument":{"version":9,"uri":"file:\
REQ--> [textDocument/completion] id=32  {"context":{"triggerKind":3},"position":{"character":13,"line":40},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/CSharp\/Po
REQ--> [textDocument/diagnostic] id=33  {"range":{"start":{"line":5,"character":0},"end":{"line":49,"character":0}},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/C
REQ--> [completionItem/resolve] id=34  {"insertTextFormat":1,"textEdit":{"newText":"private","range":{"start":{"character":8,"line":40},"end":{"character":11,"line":40}}},"kind":14,"data":{"ResultId":
<--RSP 2ms [completionItem/resolve] id=34  {"insertTextFormat":1,"data":{"ResultId":1,"TextDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/CSharp\/Portable\/Compiler\/Met
NTF--> [textDocument/didChange] {"contentChanges":[{"text":"t","range":{"start":{"character":13,"line":40},"end":{"character":13,"line":40}},"rangeLength":0}],"textDocument":{"version":10,"uri":"file:
REQ--> [textDocument/diagnostic] id=35  {"range":{"start":{"line":5,"character":0},"end":{"line":49,"character":0}},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/C
<--RSP 239ms [textDocument/completion] id=32  {"isIncomplete":true,"itemDefaults":{"data":{"TextDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/CSharp\/Portable\/Compiler
NTF--> [textDocument/didChange] {"contentChanges":[{"text":"e","range":{"start":{"character":14,"line":40},"end":{"character":14,"line":40}},"rangeLength":0}],"textDocument":{"version":11,"uri":"file:
REQ--> [textDocument/completion] id=36  {"context":{"triggerKind":3},"position":{"character":15,"line":40},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/CSharp\/Po
REQ--> [textDocument/diagnostic] id=37  {"range":{"start":{"line":5,"character":0},"end":{"line":49,"character":0}},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/C
REQ--> [completionItem/resolve] id=38  {"insertTextFormat":1,"textEdit":{"newText":"private","range":{"start":{"character":8,"line":40},"end":{"character":13,"line":40}}},"kind":14,"data":{"ResultId":
<--RSP 7ms [completionItem/resolve] id=38  {"insertTextFormat":1,"data":{"ResultId":2,"TextDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/CSharp\/Portable\/Compiler\/Met
<--RSP 230ms [textDocument/completion] id=36  {"isIncomplete":true,"itemDefaults":{"data":{"TextDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/CSharp\/Portable\/Compiler
NTF--> [textDocument/didChange] {"contentChanges":[{"text":" ","range":{"start":{"character":15,"line":40},"end":{"character":15,"line":40}},"rangeLength":0}],"textDocument":{"version":12,"uri":"file:
REQ--> [textDocument/completion] id=39  {"context":{"triggerCharacter":" ","triggerKind":2},"position":{"character":16,"line":40},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src
REQ--> [textDocument/diagnostic] id=40  {"range":{"start":{"line":5,"character":0},"end":{"line":49,"character":0}},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/C
<--RSP 167ms [textDocument/completion] id=39  null
REQ--> [textDocument/documentHighlight] id=41  {"position":{"character":16,"line":40},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/CSharp\/Portable\/Compiler\/Met
<--RSP 8ms [textDocument/documentHighlight] id=41  []
NTF--> [textDocument/didChange] {"contentChanges":[{"text":"r","range":{"start":{"character":16,"line":40},"end":{"character":16,"line":40}},"rangeLength":0}],"textDocument":{"version":13,"uri":"file:
REQ--> [textDocument/completion] id=42  {"context":{"triggerKind":1},"position":{"character":17,"line":40},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/CSharp\/Po
REQ--> [textDocument/diagnostic] id=43  {"range":{"start":{"line":5,"character":0},"end":{"line":49,"character":0}},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/C
NTF--> [textDocument/didChange] {"contentChanges":[{"text":"e","range":{"start":{"character":17,"line":40},"end":{"character":17,"line":40}},"rangeLength":0},{"text":"a","range":{"start":{"character":
REQ--> [textDocument/diagnostic] id=44  {"range":{"start":{"line":5,"character":0},"end":{"line":49,"character":0}},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/C
<--RSP 175ms [textDocument/completion] id=42  {"isIncomplete":true,"itemDefaults":{"data":{"TextDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/CSharp\/Portable\/Compiler
REQ--> [textDocument/completion] id=45  {"context":{"triggerKind":3},"position":{"character":19,"line":40},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/CSharp\/Po
NTF--> [textDocument/didChange] {"contentChanges":[{"text":"d","range":{"start":{"character":19,"line":40},"end":{"character":19,"line":40}},"rangeLength":0},{"text":"o","range":{"start":{"character":
REQ--> [textDocument/diagnostic] id=46  {"range":{"start":{"line":5,"character":0},"end":{"line":49,"character":0}},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/C
REQ--> [completionItem/resolve] id=47  {"insertTextFormat":1,"textEdit":{"newText":"readonly","range":{"start":{"character":16,"line":40},"end":{"character":17,"line":40}}},"kind":14,"data":{"ResultId
<--RSP 37ms [completionItem/resolve] id=47  {"insertTextFormat":1,"data":{"ResultId":4,"TextDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/CSharp\/Portable\/Compiler\/Me
<--RSP 267ms [textDocument/completion] id=45  {"isIncomplete":true,"itemDefaults":{"data":{"TextDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/CSharp\/Portable\/Compiler
NTF--> [textDocument/didChange] {"contentChanges":[{"text":"n","range":{"start":{"character":21,"line":40},"end":{"character":21,"line":40}},"rangeLength":0}],"textDocument":{"version":18,"uri":"file:
REQ--> [textDocument/completion] id=48  {"context":{"triggerKind":3},"position":{"character":22,"line":40},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/CSharp\/Po
REQ--> [textDocument/diagnostic] id=49  {"range":{"start":{"line":5,"character":0},"end":{"line":49,"character":0}},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/C
NTF--> [textDocument/didChange] {"contentChanges":[{"text":"l","range":{"start":{"character":22,"line":40},"end":{"character":22,"line":40}},"rangeLength":0}],"textDocument":{"version":19,"uri":"file:
REQ--> [completionItem/resolve] id=50  {"insertTextFormat":1,"textEdit":{"newText":"readonly","range":{"start":{"character":16,"line":40},"end":{"character":19,"line":40}}},"kind":14,"data":{"ResultId
REQ--> [textDocument/diagnostic] id=51  {"range":{"start":{"line":5,"character":0},"end":{"line":49,"character":0}},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/C
<--RSP 26ms [completionItem/resolve] id=50  {"insertTextFormat":1,"data":{"ResultId":5,"TextDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/CSharp\/Portable\/Compiler\/Me
NTF--> [textDocument/didChange] {"contentChanges":[{"text":"y","range":{"start":{"character":23,"line":40},"end":{"character":23,"line":40}},"rangeLength":0}],"textDocument":{"version":20,"uri":"file:
REQ--> [textDocument/diagnostic] id=52  {"range":{"start":{"line":5,"character":0},"end":{"line":49,"character":0}},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/C
<--RSP 305ms [textDocument/completion] id=48  {"isIncomplete":true,"itemDefaults":{"data":{"TextDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/CSharp\/Portable\/Compiler
NTF--> [$/cancelRequest] {"id":48}
NTF--> [textDocument/didChange] {"contentChanges":[{"text":" ","range":{"start":{"character":24,"line":40},"end":{"character":24,"line":40}},"rangeLength":0}],"textDocument":{"version":21,"uri":"file:
REQ--> [textDocument/completion] id=53  {"context":{"triggerCharacter":" ","triggerKind":2},"position":{"character":25,"line":40},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src
REQ--> [textDocument/diagnostic] id=54  {"range":{"start":{"line":5,"character":0},"end":{"line":49,"character":0}},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/C
<--RSP 160ms [textDocument/completion] id=53  null
REQ--> [textDocument/documentHighlight] id=55  {"position":{"character":25,"line":40},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/CSharp\/Portable\/Compiler\/Met
<--RSP 6ms [textDocument/documentHighlight] id=55  []
NTF--> [textDocument/didChange] {"contentChanges":[{"text":"x","range":{"start":{"character":25,"line":40},"end":{"character":25,"line":40}},"rangeLength":0}],"textDocument":{"version":22,"uri":"file:
REQ--> [textDocument/completion] id=56  {"context":{"triggerKind":1},"position":{"character":26,"line":40},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/CSharp\/Po
REQ--> [textDocument/diagnostic] id=57  {"range":{"start":{"line":5,"character":0},"end":{"line":49,"character":0}},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/C
<--RSP 148ms [textDocument/completion] id=56  {"isIncomplete":true,"itemDefaults":{"data":{"TextDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/CSharp\/Portable\/Compiler
NTF--> [textDocument/didChange] {"contentChanges":[{"text":" ","range":{"start":{"character":26,"line":40},"end":{"character":26,"line":40}},"rangeLength":0}],"textDocument":{"version":23,"uri":"file:
REQ--> [textDocument/completion] id=58  {"context":{"triggerCharacter":" ","triggerKind":2},"position":{"character":27,"line":40},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src
REQ--> [textDocument/diagnostic] id=59  {"range":{"start":{"line":5,"character":0},"end":{"line":49,"character":0}},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/C
<--RSP 91ms [textDocument/completion] id=58  null
<--RSP 4103ms [textDocument/diagnostic] id=23  {"kind":"full","resultId":"PublicDocumentPullDiagnosticsHandler(category: ):1","items":[{"range":{"start":{"character":16,"line":268},"end":{"character":
REQ--> [textDocument/documentHighlight] id=60  {"position":{"character":27,"line":40},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/CSharp\/Portable\/Compiler\/Met
<--RSP 5ms [textDocument/documentHighlight] id=60  []
NTF--> [textDocument/didChange] {"contentChanges":[{"text":"s","range":{"start":{"character":27,"line":40},"end":{"character":27,"line":40}},"rangeLength":0}],"textDocument":{"version":24,"uri":"file:
REQ--> [textDocument/completion] id=61  {"context":{"triggerKind":1},"position":{"character":28,"line":40},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/CSharp\/Po
REQ--> [textDocument/diagnostic] id=62  {"range":{"start":{"line":5,"character":0},"end":{"line":49,"character":0}},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/C
NTF--> [textDocument/didChange] {"contentChanges":[{"text":"t","range":{"start":{"character":28,"line":40},"end":{"character":28,"line":40}},"rangeLength":0}],"textDocument":{"version":25,"uri":"file:
REQ--> [textDocument/diagnostic] id=63  {"range":{"start":{"line":5,"character":0},"end":{"line":49,"character":0}},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/C
<--RSP 234ms [textDocument/completion] id=61  null
NTF--> [textDocument/didChange] {"contentChanges":[{"text":"r","range":{"start":{"character":29,"line":40},"end":{"character":29,"line":40}},"rangeLength":0}],"textDocument":{"version":26,"uri":"file:
REQ--> [textDocument/diagnostic] id=64  {"range":{"start":{"line":5,"character":0},"end":{"line":49,"character":0}},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/C
NTF--> [textDocument/didChange] {"contentChanges":[{"text":"i","range":{"start":{"character":30,"line":40},"end":{"character":30,"line":40}},"rangeLength":0},{"text":"n","range":{"start":{"character":
REQ--> [textDocument/diagnostic] id=65  {"range":{"start":{"line":5,"character":0},"end":{"line":49,"character":0}},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/C
NTF--> [textDocument/didChange] {"contentChanges":[{"text":"g","range":{"start":{"character":32,"line":40},"end":{"character":32,"line":40}},"rangeLength":0},{"text":";","range":{"start":{"character":
REQ--> [textDocument/diagnostic] id=66  {"range":{"start":{"line":5,"character":0},"end":{"line":49,"character":0}},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Compilers\/C
**RESPONSE TO FIRST DIAGNOSTIC AFTER 5s**<--RSP 5166ms [textDocument/diagnostic] id=26  {"kind":"full","resultId":"PublicDocumentPullDiagnosticsHandler(category: ):2","items":[{"range":{"start":{"character":16,"line":268},"end":{"character":
```

vscode

```
NTF--> [textDocument/didChange] {"contentChanges":[{"range":{"start":{"character":8,"line":44},"end":{"character":8,"line":44}},"rangeLength":0,"text":"p"}],"textDocument":{"uri":"file:\/\/\/Users\/mp
NTF--> [$/cancelRequest] {"id":94}
NTF--> [$/cancelRequest] {"id":95}
NTF--> [$/cancelRequest] {"id":96}
REQ--> [textDocument/diagnostic] id=97  {"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Tools\/ExternalAccess\/Razor\/RazorDynamicFileInfo.cs"},"identifier":"NonLocal"}
REQ--> [textDocument/diagnostic] id=98  {"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Tools\/ExternalAccess\/Razor\/RazorDynamicFileInfo.cs"},"previousResultId":"PublicDocum
REQ--> [textDocument/diagnostic] id=99  {"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Tools\/ExternalAccess\/Razor\/RazorDynamicFileInfo.cs"},"previousResultId":"PublicDocum
REQ--> [textDocument/diagnostic] id=100  {"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Tools\/ExternalAccess\/Razor\/RazorDynamicFileInfo.cs"},"previousResultId":"PublicDocu
REQ--> [textDocument/diagnostic] id=101  {"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Tools\/ExternalAccess\/Razor\/RazorDynamicFileInfo.cs"},"identifier":"XamlDiagnostics"
REQ--> [textDocument/diagnostic] id=102  {"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Tools\/ExternalAccess\/Razor\/RazorDynamicFileInfo.cs"},"previousResultId":"PublicDocu
REQ--> [textDocument/diagnostic] id=103  {"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Tools\/ExternalAccess\/Razor\/RazorDynamicFileInfo.cs"},"identifier":"HotReloadDiagnos
REQ--> [textDocument/diagnostic] id=104  {"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Tools\/ExternalAccess\/Razor\/RazorDynamicFileInfo.cs"},"previousResultId":"PublicDocu
REQ--> [textDocument/diagnostic] id=105  {"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Tools\/ExternalAccess\/Razor\/RazorDynamicFileInfo.cs"},"identifier":"WorkspaceDocumen
<--ERR [nil] id=94  {"message":"The task was cancelled.","code":-32800}
<--RSP 2ms [textDocument/diagnostic] id=97  null
<--ERR [nil] id=95  {"message":"The task was cancelled.","code":-32800}
<--RSP 0ms [textDocument/diagnostic] id=105  null
<--RSP 1ms [textDocument/diagnostic] id=101  null
<--RSP 1ms [textDocument/diagnostic] id=102  {"resultId":"PublicDocumentPullDiagnosticsHandler(category: enc):1","kind":"unchanged"}
<--RSP 1ms [textDocument/diagnostic] id=103  null
<--ERR [nil] id=96  {"message":"The task was cancelled.","code":-32800}
<--RSP 4ms [textDocument/diagnostic] id=104  {"resultId":"PublicDocumentPullDiagnosticsHandler(category: syntax):1","kind":"unchanged"}
<--RSP 7ms [textDocument/diagnostic] id=98  {"resultId":"PublicDocumentPullDiagnosticsHandler(category: DocumentAnalyzerSyntax):1","kind":"unchanged"}
<--RSP 5ms [textDocument/diagnostic] id=100  {"resultId":"PublicDocumentPullDiagnosticsHandler(category: DocumentCompilerSemantic):1","kind":"unchanged"}
<--RSP 7ms [textDocument/diagnostic] id=99  {"resultId":"PublicDocumentPullDiagnosticsHandler(category: DocumentAnalyzerSemantic):1","kind":"unchanged"}
REQ--> [textDocument/completion] id=106  {"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Tools\/ExternalAccess\/Razor\/RazorDynamicFileInfo.cs"},"position":{"character":9,"lin
REQ--> [textDocument/documentHighlight] id=107  {"position":{"character":9,"line":44},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Tools\/ExternalAccess\/Razor\/RazorDynamic
<--RSP 25ms [textDocument/documentHighlight] id=107  []
REQ--> [textDocument/_vs_getProjectContexts] id=108  {"_vs_textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Tools\/ExternalAccess\/Razor\/RazorDynamicFileInfo.cs"}}
REQ--> [textDocument/inlayHint] id=109  {"range":{"start":{"character":0,"line":0},"end":{"character":0,"line":57}},"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/roslyn\/src\/Tools\/Exter
<--RSP 2ms [textDocument/_vs_getProjectContexts] id=108  {"_vs_defaultIndex":0,"_vs_projectContexts":[{"_vs_id":"7a2dea60-2780-40c5-815b-3a8d3d9ada37|LspMiscellaneousFilesWorkspace Files Project for \
<--RSP 5ms [textDocument/inlayHint] id=109  []
NTF--> [textDocument/didChange] {"contentChanges":[{"range":{"start":{"character":9,"line":44},"end":{"character":9,"line":44}},"rangeLength":0,"text":"r"}],"textDocument":{"uri":"file:\/\/\/Users\/mp

```

**Comment by molostovvs** [comment]
There is already some work being done on pull diagnostics request only after the file is saved https://github.com/neovim/neovim/issues/29792.

But it seems that the problem here is really that the `textDocument/diagnostic` lacks additional information, as @mparq pointed out

**Comment by molostovvs** [comment]
yep, it seems the `textDocument/diagnostic` [implementation in neovim](https://github.com/neovim/neovim/blob/8f84167c30692555d3332565605e8a625aebc43c/runtime/lua/vim/lsp/diagnostic.lua?plain=1#L377) does not support the `previousResultId` and `identifier` parameters [that are in the lsp spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#documentDiagnosticParams).

<details>
<summary>note to self</summary>

read more code of the [current implementation of pull diagnostics](https://github.com/dotnet/roslyn/blob/main/src/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs#L171) in roslyn

</details>

**Comment by seblj** [comment]
Yeah okay, so it should probably be implemented in neovim then. I cannot guarantee that I myself will have time to do it, but I am not opposed to try and implement it if I find some time for it.

If anyone beats me to it, then great😄

**Comment by Wordluc** [comment]
I've implemented a change like this within `function Client:_request(method, params, handler, bufnr)`:
```
  if last_result ~= nil then
    if last_result.resultId ~= nil then
      params.previousResultId = last_result.resultId
    end
  end
```
With this modification, the responses are significantly quicker. I haven't committed the change on a fork (for link the commit) because I'm not on Linux and am unable to build on Windows. This is a temporary fix.

**Comment by seblj** [comment]
So I am thinking about overriding the client request method and keeping the previous result id to see if that can make it better, but I don't know if I am just tired or something, but I was sure that this was going to work, and I cannot see what is wrong with this:
```lua
    config.on_attach = function(client, _)
        local req_inner = client.request

        function client:request(method, params, handler, req_bufnr)
            if method == "textDocument/diagnostic" and last_result ~= nil and last_result.resultId ~= nil then
                params.previousResultId = last_result.resultId
            end

            if handler then
                return req_inner(self, method, params, function(err, result, context)
                    last_result = result
                    return handler(err, result, context)
                end, req_bufnr)
            else
                return req_inner(self, method, params, handler, req_bufnr)
            end
        end
    end
```

If anyone can see what is wrong, then please give me a headsup😅 Maybe I am just tired or something

**Comment by drzbida** [comment]
I made it work directly in the plugin. I'll open a PR after i check one more thing

**Comment by seblj** [comment]
Could someone who has a good reproduction of slowness test this: https://github.com/seblj/roslyn.nvim/pull/105

I was hoping it would sort of fix the lagging with diagnostics for me like here, but it seems to become faster with some time. That also seems to be the case without the fix though. I don't really have a good reproducible myself, but if this makes it faster for some of you we should of course do this


https://github.com/user-attachments/assets/adc4f394-0fdf-4eaf-bd0b-aa8fb86b2faa

**Comment by mparq** [comment]
@drzbida I gave your fix a try. Things do seem faster now thank you :) especially important for me is when editing for a while I'm not seeing the same behavior where nvim just freezes. That said, I do still see the diagnostic "backup" behavior, maybe that will need to be fixed using "identifier" in https://github.com/neovim/neovim/issues/31454. This is what I'm seeing in the traces now on the razor project.

because the full diagnostic pull for large projects still takes much longer to return than it takes to type, we get the issue where many in flight requests won't have previousResultId or will have an outdated one.

truncated trace (removed non diagnostic and didChange requests)
```
REQ--> [textDocument/diagnostic] id=16  {"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/razor\/src\/Razor\/src\/Microsoft.A
REQ--> [textDocument/diagnostic] id=19  {"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/razor\/src\/Razor\/src\/Microsoft.A
NTF--> [textDocument/didChange] {"contentChanges":[{"range":{"end":{"character":5,"line":27},"start":{"character":5,"line":27}},"text":
REQ--> [textDocument/diagnostic] id=20  {"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/razor\/src\/Razor\/src\/Microsoft.A
NTF--> [textDocument/didChange] {"contentChanges":[{"range":{"end":{"character":6,"line":27},"start":{"character":6,"line":27}},"text":
REQ--> [textDocument/diagnostic] id=22  {"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/razor\/src\/Razor\/src\/Microsoft.A
NTF--> [textDocument/didChange] {"contentChanges":[{"range":{"end":{"character":7,"line":27},"start":{"character":7,"line":27}},"text":
REQ--> [textDocument/diagnostic] id=26  {"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/razor\/src\/Razor\/src\/Microsoft.A
NTF--> [textDocument/didChange] {"contentChanges":[{"range":{"end":{"character":8,"line":27},"start":{"character":8,"line":27}},"text":
REQ--> [textDocument/diagnostic] id=29  {"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/razor\/src\/Razor\/src\/Microsoft.A
NTF--> [textDocument/didChange] {"contentChanges":[{"range":{"end":{"character":9,"line":27},"start":{"character":9,"line":27}},"text":
REQ--> [textDocument/diagnostic] id=31  {"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/razor\/src\/Razor\/src\/Microsoft.A
<--RSP 3444ms [textDocument/diagnostic] id=14  {"kind":"full","items":[],"resultId":"PublicDocumentPullDiagnosticsHandler(category: ):1
NTF--> [textDocument/didChange] {"contentChanges":[{"range":{"end":{"character":10,"line":27},"start":{"character":10,"line":27}},"text
REQ--> [textDocument/diagnostic] id=34  {"previousResultId":"PublicDocumentPullDiagnosticsHandler(category: ):1","range":{"end":{"chara
NTF--> [textDocument/didChange] {"contentChanges":[{"range":{"end":{"character":11,"line":27},"start":{"character":11,"line":27}},"text
REQ--> [textDocument/diagnostic] id=37  {"previousResultId":"PublicDocumentPullDiagnosticsHandler(category: ):1","range":{"end":{"chara
NTF--> [textDocument/didChange] {"contentChanges":[{"range":{"end":{"character":12,"line":27},"start":{"character":12,"line":27}},"text
REQ--> [textDocument/diagnostic] id=40  {"previousResultId":"PublicDocumentPullDiagnosticsHandler(category: ):1","range":{"end":{"chara
<--RSP 1609ms [textDocument/diagnostic] id=16  {"kind":"full","items":[],"resultId":"PublicDocumentPullDiagnosticsHandler(category: ):2
NTF--> [textDocument/didChange] {"contentChanges":[{"range":{"end":{"character":13,"line":27},"start":{"character":13,"line":27}},"text
REQ--> [textDocument/diagnostic] id=42  {"previousResultId":"PublicDocumentPullDiagnosticsHandler(category: ):2","range":{"end":{"chara
```

**Comment by drzbida** [comment]
> Could someone who has a good reproduction of slowness test this: #105
> 
> I was hoping it would sort of fix the lagging with diagnostics for me like here, but it seems to become faster with some time. That also seems to be the case without the fix though. I don't really have a good reproducible myself, but if this makes it faster for some of you we should of course do this
> Screen.Recording.2024-12-05.at.22.08.38.mov

Wow, I never imagined it to be that slow. In any case, an issue should definitely be opened in neovim itself as these parameters are part of the [3.17 LSP spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#documentDiagnosticParams) and they are actually used in roslyn for some optimizations 

Maybe it would be useful if you could try a Windows VM and opening your project to see if it behaves the same? (Unless someone here using Windows is having these issues). I've personally never seen this and we have a pretty big monorepo at my company. If it doesn't happen then it must be an issue with Roslyn (but that would mean you should see this issue in vscode too) or Neovim itself.

There could be another way it doesn't happen in vscode, Microsoft sometimes have modified versions of various tools in vscode. Although in all cases I know of they are clearly different in name, it takes 2 minutes to test using the dll from vscode, who knows.. 

You can find it here

C:\Users\bida\.vscode\extensions\ms-dotnettools.csharp-2.55.29-win32-x64\.roslyn\Microsoft.CodeAnalysis.LanguageServer.dll

For linux it's probably something like

~/.config/.vscode/extensions/ms-dotnettools.csharp-*/.roslyn/Microsoft.CodeAnalysis.LanguageServer.dll

**Comment by drzbida** [comment]
> @drzbida I gave your fix a try. Things do seem faster now thank you :) especially important for me is when editing for a while I'm not seeing the same behavior where nvim just freezes. That said, I do still see the diagnostic "backup" behavior, maybe that will need to be fixed using "identifier" in [neovim/neovim#31454](https://github.com/neovim/neovim/issues/31454). This is what I'm seeing in the traces now on the razor project. (I've been using the file Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/Legacy/CSharpCodeParser.cs to test as it has 3k+ lines)
> 
> because the full diagnostic pull for large projects still takes much longer to return than it takes to type, we get the issue where many in flight requests won't have previousResultId or will have an outdated one.
> 
> truncated trace (removed non diagnostic and didChange requests)
> 
> ```
> REQ--> [textDocument/diagnostic] id=16  {"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/razor\/src\/Razor\/src\/Microsoft.A
> REQ--> [textDocument/diagnostic] id=19  {"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/razor\/src\/Razor\/src\/Microsoft.A
> NTF--> [textDocument/didChange] {"contentChanges":[{"range":{"end":{"character":5,"line":27},"start":{"character":5,"line":27}},"text":
> REQ--> [textDocument/diagnostic] id=20  {"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/razor\/src\/Razor\/src\/Microsoft.A
> NTF--> [textDocument/didChange] {"contentChanges":[{"range":{"end":{"character":6,"line":27},"start":{"character":6,"line":27}},"text":
> REQ--> [textDocument/diagnostic] id=22  {"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/razor\/src\/Razor\/src\/Microsoft.A
> NTF--> [textDocument/didChange] {"contentChanges":[{"range":{"end":{"character":7,"line":27},"start":{"character":7,"line":27}},"text":
> REQ--> [textDocument/diagnostic] id=26  {"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/razor\/src\/Razor\/src\/Microsoft.A
> NTF--> [textDocument/didChange] {"contentChanges":[{"range":{"end":{"character":8,"line":27},"start":{"character":8,"line":27}},"text":
> REQ--> [textDocument/diagnostic] id=29  {"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/razor\/src\/Razor\/src\/Microsoft.A
> NTF--> [textDocument/didChange] {"contentChanges":[{"range":{"end":{"character":9,"line":27},"start":{"character":9,"line":27}},"text":
> REQ--> [textDocument/diagnostic] id=31  {"textDocument":{"uri":"file:\/\/\/Users\/mparq\/code\/ext\/razor\/src\/Razor\/src\/Microsoft.A
> <--RSP 3444ms [textDocument/diagnostic] id=14  {"kind":"full","items":[],"resultId":"PublicDocumentPullDiagnosticsHandler(category: ):1
> NTF--> [textDocument/didChange] {"contentChanges":[{"range":{"end":{"character":10,"line":27},"start":{"character":10,"line":27}},"text
> REQ--> [textDocument/diagnostic] id=34  {"previousResultId":"PublicDocumentPullDiagnosticsHandler(category: ):1","range":{"end":{"chara
> NTF--> [textDocument/didChange] {"contentChanges":[{"range":{"end":{"character":11,"line":27},"start":{"character":11,"line":27}},"text
> REQ--> [textDocument/diagnostic] id=37  {"previousResultId":"PublicDocumentPullDiagnosticsHandler(category: ):1","range":{"end":{"chara
> NTF--> [textDocument/didChange] {"contentChanges":[{"range":{"end":{"character":12,"line":27},"start":{"character":12,"line":27}},"text
> REQ--> [textDocument/diagnostic] id=40  {"previousResultId":"PublicDocumentPullDiagnosticsHandler(category: ):1","range":{"end":{"chara
> <--RSP 1609ms [textDocument/diagnostic] id=16  {"kind":"full","items":[],"resultId":"PublicDocumentPullDiagnosticsHandler(category: ):2
> NTF--> [textDocument/didChange] {"contentChanges":[{"range":{"end":{"character":13,"line":27},"start":{"character":13,"line":27}},"text
> REQ--> [textDocument/diagnostic] id=42  {"previousResultId":"PublicDocumentPullDiagnosticsHandler(category: ):2","range":{"end":{"chara
> ```

I'm not sure identifier will help in performance. I wrote more in the PR. Someone else also looking through the linked code would definitely help

**Comment by seblj** [comment]
Hmm, I can see if I can test this on linux and windows as well. That was tested on my mac, and that is also a very small project as well with around 50 files or something, and none of them over 200-300 lines long I think

**Comment by drzbida** [comment]
I'm finally managing to reproduce in on some huge files. 

Something very weird (imo) is that neovim sends a textDocument/diagnostic for every single character you've pressed after exiting normal mode. Default behavior is to update diagnostics on InsertLeave

```
• {update_in_insert}?  (`boolean`, default: `false`) Update diagnostics
                       in Insert mode (if `false`, diagnostics are
                       updated on |InsertLeave|)
```

I expected it to be done in a single request to be honest, but I guess keeping track of the changes is hard since you can also move with the arrow keys / etc ? It's definitely possible to do in a single request, after executing a command like `:g/static/d` requests diagnostics only once even though it deleted 217 lines for me. Could be worth exploring if an issue for something like this exists in neovim.. 

But then again, vscode updates on every keystroke too and works just fine, so there must be something else

**Comment by seblj** [comment]
> Something very weird (imo) is that neovim sends a textDocument/diagnostic for every single character you've pressed after exiting insert mode

So it doesn't actually send it on TextChangedI/TextChangedP? But in vscode it does since that doesn't have a concept of different modes?

Do you notice any difference with/without the PR?

**Comment by molostovvs** [comment]
In my case (I found a 6k line test file) the changes from https://github.com/seblj/roslyn.nvim/pull/105 didn't make a significant difference, unfortunately

The vscode for the same change in the file needs ~5 seconds to display the diagnostics correctly.

<details>
<summary><i>Here are 2 screen recordings for comparison: from main branch and from PR</i></summary>

https://github.com/user-attachments/assets/5c44944d-29f5-46fd-8929-ebe58a85f84f

https://github.com/user-attachments/assets/13b11114-dd19-494e-a939-0b8eb1ec8156
</details>

**Comment by seblj** [comment]
> So it doesn't actually send it on TextChangedI/TextChangedP? But in vscode it does since that doesn't have a concept of different modes?

No, I tested this and it seems like it sends the requests while I am typing

**Comment by seblj** [comment]
I tried to change it out to only update on `InsertLeave` by modifying neovim source code directly, and it seems significantly faster than the video I posted earlier in this thread...

It was a bit weird though, because it didn't clear the diagnostics again after undo and such since I only refreshed them on InsertLeave

**Comment by tris203** [comment]
My other thought is that it is an issue with inflight requests not being cancelled when they become outdated.

That would make the request feel snappier, if at the end of typing you didn't have to wait for all the outdated partially typed values to be computed.

**Comment by drzbida** [comment]
Having a debounce for the diagnostics would help a lot. To me it seems like it's done in vscode too (at least with my config, where I don't remember changing anything like this)

I threw together a quick test hack for 0.10 only

```lua
local debounce_timer = nil

config.on_attach = function(client, attach_bufnr)
    local original_request = client.request
    local last_diagnostic = nil
    local pending_params = nil

    local function send_diagnostic_request()
        if pending_params then
            local params = pending_params
            pending_params = nil
            original_request("textDocument/diagnostic", params, function(err, result, ctx)
                print("actually sending diagnostic request")
                if result and result.resultId then
                    last_diagnostic = result
                end
                local handler = roslyn_config.config.handlers["textDocument/diagnostic"]
                    or vim.lsp.handlers["textDocument/diagnostic"]
                handler(err, result, ctx)
            end)
        end
    end

    client.request = function(method, params, handler, req_bufnr)
        if method ~= "textDocument/diagnostic" then
            return original_request(method, params, handler, req_bufnr)
        end

        params.previousResultId = last_diagnostic and last_diagnostic.resultId
        pending_params = params

        if debounce_timer then
            debounce_timer:stop()
        end

        debounce_timer = vim.defer_fn(send_diagnostic_request, 300)
    end

    if roslyn_config.config.on_attach then
        roslyn_config.config.on_attach(client, attach_bufnr)
    end
end
```

But i really am not sure why this is needed as this should work from my understanding with the debounce_text_changes flag in vim.lsp.client.flags

```
• {flags}                 (`table`) A table with flags for the client.
                          The current (experimental) flags are:
                          • {allow_incremental_sync}? (`boolean`,
                            default: `true`) Allow using incremental
                            sync for buffer edits
                          • {debounce_text_changes} (`integer`, default:
                            `150`) Debounce `didChange` notifications to
                            the server by the given number in
                            milliseconds. No debounce occurs if `nil`.
                          • {exit_timeout} (`integer|false`, default:
                            `false`) Milliseconds to wait for server to
                            exit cleanly after sending the "shutdown"
                            request before sending kill -15. If set to
                            false, nvim exits immediately after sending
                            the "shutdown" request to the server.
```

But it has no effect for me in roslyn. Is it possible it was not adapted for pull diagnostics?

**Comment by drzbida** [comment]
To summarize, so far I see these problems which are not related to each other:

- Diagnostics are requested in insert mode even with default settings when they are not shown (Neovim)
- Debounce text changes has no effect for pull diagnostics ? (Neovim)
- ~~We don't send previousResultId for pull diagnostics~~
    - It's still slow, unclear how much it helped (maybe @molostovvs can also try with lsp references disabled) (Reason unknown)

**Comment by tris203** [comment]
I have tried a potential fix in neovim. Can one of you guys that can reproduce this try the PR?

**Comment by seblj** [comment]
For me it seems MUCH better just in this small file I have with under 100 lines of code. See the video I posted earlier for comparison

https://github.com/user-attachments/assets/9c1a6b17-3cac-4cd1-8b7b-8db817a1213c

**Comment by Wordluc** [comment]
same thing, it's much better, would be possible to do the same stuff in the plugin? thanks https://github.com/seblj/roslyn.nvim/pull/105 we have a wrapper for the request

**Comment by Wordluc** [comment]
I tried to cancell the request inside the plugin and the result was really good(2000 lines of code).
before:
https://github.com/user-attachments/assets/9658d023-91d7-4c7d-9bf2-9f5517845a3a
after:
https://github.com/user-attachments/assets/48d1a5e4-8368-4574-85c0-f99dcf4e3dd4
code that use https://github.com/seblj/roslyn.nvim/pull/105
```
	client.request = function(method, params, handler, req_bufnr)
				--delete request
				for rid, req in pairs(client.requests) do
					if req.method == method and req.type == 'pending' and req.bufnr == bufnr then
						client.cancel_request(rid)
					end
				end
				----
.....ecc

```

**Comment by tris203** [comment]
I don't know @seblj's thoughts but I think the best course of action is to get it fixed upstream and backported to 0.10.x

**Comment by seblj** [comment]
Yeah, I would prefer if we could get it merged and backported.

If it's not being backported, then I could be open to add a hack only for 0.10 since it's such a big improvement in performance.

However, I would like to wait with any hack until something is eventually merged into neovim

**Comment by tris203** [comment]
The range is because logic is shared within Neovims diagnostic and inlayhint refresh implementations
It isn't strictly correct, but it is what it is at the moment.

There is also no central LSP datastore to hold the identifier information, so as a workaround for "one shot" diagnostics which is all that Neovim supports _at the moment_. Then it seems the best workaround.

It might be good for you to create an issue in neovim/neovim for that range being included as that is something that can be corrected

**Comment by seblj** [comment]
Okay, it was merged in neovim now, and a backport seems to be in progress

**Comment by mparq** [comment]
Just piling on here, just got around to testing. Things are way faster now! Thanks @tris203. I'm also seeing that with the cancelRequest changes, the lsp textDocumemt/diagnostic responses from the language server consistently return in ~150-250ms for me, whereas before the changes they could take up to 5000ms which I'm now thinking was scaling based on how fast I was typing. Still not the ~5ms responses from vscode diagnostic requests which include identifier field but so much better than it was

**Comment by Sverre-W** [comment]
I'm not sure about 10.2, but the lagging experience broke my workflow so after the fix I switched to 11 and since then I have no issues with lag anymore.

**Comment by Radu-C-Martin** [comment]
Not working of me yet on 10.2, but updating to 11 does siiiiiignificantly improve it.

**Comment by seblj** [comment]
I guess you need to reinstall 0.10.2 to notice any improvements

**Comment by ilyasakin** [comment]
I'm guessing the problem was originated from either neovim itself or roslyn. Just updated my neovim to latest commit and upgraded roslyn to latest version from Mason. 

Diagnostics are WAAAAAAAY faster now.
